pythonwhat/tasks.py

```diff
@@ -321,14 +321,21 @@ def taskRunEval(tree,
             tree = ast.Expression(tree)
 
         # Expression code takes precedence over tree code
-        if expr_code: code = expr_code
-        else:         code = compile(tree, "<script>", mode)
+        if expr_code:
+            code = expr_code
+            tree = ast.parse(code, mode=mode)
+        else:
+            code = compile(tree, "<script>", mode)
 
         # Set up environment --------------------------------------------------
         # avoid deepy copy if specified, or just looking up variable by name
+        if isinstance(tree, ast.Expression):
```

Comment:
I think I missed something, because I can't tell how these changes "Limit impact of unpicklable object" (but there are tests so they must do it?!)

Reply:
It unpacks the expression so in the if-else below in more cases the if branch, without deep copy, is used.

Reply:
Ah, yeah--I see. It looks like you are parsing "eval" style code to an AST expression, and replacing `tree` in the function with that.

It seems that before, there was a bug, where it used the state tree to determine whether to copy (as if executing it), even if you passed custom eval code using `expr_code`.

If expr_code is used a lot seems like it could save a lot of time!